# Phase A — Data Preparation

This notebook performs initial data loading, target-feature separation, and feature classification for the Ames Housing dataset.

---
## Tasks Overview
- **Task 1**: Load dataset & verify `SalePrice` and shape.
- **Task 2**: Separate target (`y = SalePrice`) and 80 candidate features (`X`).
- **Task 3**: Identify feature types (Numerical, Nominal Categorical, Ordinal Categorical) without dropping features.

### Task 1 — Load the Dataset

Load `dataset.csv` into Pandas, confirm `SalePrice` exists, and check dataset dimensions.

In [9]:
import os
import pandas as pd
import numpy as np

# Load dataset
df = pd.read_csv("dataset.csv")

# Confirm presence of SalePrice
has_sale_price = "SalePrice" in df.columns
print(f"Contains 'SalePrice': {has_sale_price}")

# Check shape
print(f"Dataset shape: {df.shape}")
df.head()

Contains 'SalePrice': True
Dataset shape: (1460, 81)


,Id,MSSubClass,MSZoning,LotFrontage,LotArea,Street,Alley,LotShape,LandContour,Utilities,...,PoolArea,PoolQC,Fence,MiscFeature,MiscVal,MoSold,YrSold,SaleType,SaleCondition,SalePrice
0,1,60,RL,65.0,8450,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,2,2008,WD,Normal,208500
1,2,20,RL,80.0,9600,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,5,2007,WD,Normal,181500
2,3,60,RL,68.0,11250,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,9,2008,WD,Normal,223500
3,4,70,RL,60.0,9550,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,2,2006,WD,Abnorml,140000
4,5,60,RL,84.0,14260,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,12,2008,WD,Normal,250000


### Task 1 Analysis & Verification
- **Dataset Shape**: 1,460 rows and 81 columns.
- **Target Verification**: `SalePrice` column is present and confirmed as target column.

### Task 2 — Separate Target and Features

Separate the target variable (`y`) from the candidate feature set (`X`).
- `X` = `df.drop("SalePrice", axis=1)` (80 candidate features)
- `y` = `df["SalePrice"]` (target series)

In [ ]:
# Separate features and target
X = df.drop("SalePrice", axis=1)
y = df["SalePrice"]

print(f"X shape (candidate features): {X.shape}")
print(f"y shape (target): {y.shape}")

### Task 2 Analysis & Verification
- **Feature Set (`X`)**: Contains 80 candidate features across 1,460 observations.
- **Target (`y`)**: Contains 1,460 continuous target values for `SalePrice`.

### Task 3 — Identify Feature Types

Separate the 80 candidate features in `X` into **Numerical**, **Nominal Categorical**, and **Ordinal Categorical** features.
All 80 candidate features are retained.

In [ ]:
# 1. Define Ordinal Categorical Features (features with intrinsic ordering/ratings)
ordinal_features = [
    'LotShape', 'Utilities', 'LandSlope', 'ExterQual', 'ExterCond', 
    'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2',
    'HeatingQC', 'Electrical', 'KitchenQual', 'Functional', 'FireplaceQu',
    'GarageFinish', 'GarageQual', 'GarageCond', 'PavedDrive', 'PoolQC', 'Fence'
]

# 2. Extract string/object categorical columns from X
object_cols = X.select_dtypes(include=['object', 'string']).columns.tolist()

# 3. Define Nominal Categorical Features (unordered string categories)
nominal_features = [col for col in object_cols if col not in ordinal_features]

# 4. Define Numerical Features (continuous, discrete, and numeric-encoded variables)
numerical_features = X.select_dtypes(include=['int64', 'float64']).columns.tolist()

print(f"Numerical Features ({len(numerical_features)}):")
print(numerical_features)
print(f"\nNominal Categorical Features ({len(nominal_features)}):")
print(nominal_features)
print(f"\nOrdinal Categorical Features ({len(ordinal_features)}):")
print(ordinal_features)

# Sanity check total feature count
total_classified = len(numerical_features) + len(nominal_features) + len(ordinal_features)
print(f"\nTotal Features Classified: {total_classified} / {X.shape[1]}")

### Task 3 Analysis & Summary

The 80 candidate features in `X` are categorized as follows:

1. **Numerical Features (37)**:
   - Includes continuous measurements (e.g., `LotFrontage`, `LotArea`, `1stFlrSF`, `GrLivArea`, `GarageArea`), discrete counts (e.g., `FullBath`, `BedroomAbvGr`), year columns (`YearBuilt`, `YearRemodAdd`, `GarageYrBlt`, `YrSold`), identifiers (`Id`), as well as numeric-encoded features (`MSSubClass`, `OverallQual`, `OverallCond`).

2. **Nominal Categorical Features (22)**:
   - Unordered categories: `MSZoning`, `Street`, `Alley`, `LandContour`, `LotConfig`, `Neighborhood`, `Condition1`, `Condition2`, `BldgType`, `HouseStyle`, `RoofStyle`, `RoofMatl`, `Exterior1st`, `Exterior2nd`, `MasVnrType`, `Foundation`, `Heating`, `CentralAir`, `GarageType`, `MiscFeature`, `SaleType`, `SaleCondition`.

3. **Ordinal Categorical Features (21)**:
   - Categorical features with inherent order (ratings/quality scales): `LotShape`, `Utilities`, `LandSlope`, `ExterQual`, `ExterCond`, `BsmtQual`, `BsmtCond`, `BsmtExposure`, `BsmtFinType1`, `BsmtFinType2`, `HeatingQC`, `Electrical`, `KitchenQual`, `Functional`, `FireplaceQu`, `GarageFinish`, `GarageQual`, `GarageCond`, `PavedDrive`, `PoolQC`, `Fence`.

> [!NOTE]
> All 80 candidate features are preserved for subsequent feature engineering and encoding phases.